server/server/data_loader/fuzzymatcher.py

```diff
@@ -0,0 +1,68 @@
+from collections import defaultdict, Counter
+
+class FuzzyMatcher:
+    """
+    
+    Uses ngrams to identify similiar words
+    
+    You'd think this would be a solved problem but I looked into other
+    options on the internet and there are a lot but most of them
+    are either O(N^2) or just way too slow.
+    This method works surprisingly well for pali and takes mere seconds
+    to find similiarties within a 40000 word dictionary.
+    
+    """
+    
+    # Future Improvements that aren't a huge deal but would be nice:
+    # 1. Certain combinations like kh, dh etc are actually a single
+    #    letter and should be treated as such.
+    
+    def __init__(self, words, threshold=0.0):
+        """
+        Add words to be matched against
+        
+        If threshold is 0.01 than any ngram which appears in more than
+        1% of words will be ignored, essentially like a stopword"""
+        self.words = words
+        self.ngram_mapping = defaultdict(set)
+        for word in words:
+            self.add(word)
+            
+        if threshold > 0:
+            self._prune_common(len(words) * threshold)
+    
+    def _prune_common(self, max_count):
+        "Prune any ngrams that appear in more than max_count words"
+        for k, v in list(self.ngram_mapping.items()):
+            if len(v) > max_count:
+                del self.ngram_mapping[k]                    
+        
+    def make_ngrams(self, word, length=4):
```

Comment:
It can be turned into generator

Reply:
It returns a set... but I'll rename it to _make_ngrams since the return type is peculiar to the form I want to use in the other methods.